src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1989,6 +1989,9 @@ internal UserAction StartUserAction(CancellationToken cancellationToken = defaul
 
             lock (this)
             {
+                if (IsBroken)
```

Comment:
Technically, the same bug can occur because the connector was cleanly closed (not broken). I think this should only be possible if the user violates our contract and (accidentally) closes the connection from one thread, while executing a query in the other. However, since we want to improve concurrency detection, we can have a switch here and check for Closed as well.

Also, it may be a good idea to not just throw _breakReason, but to prepend it with a message such as "The connection was previously broken because of the following: ", just to provide a bit more info on what's going on (it would help us in future bug reports :wink: )

And with the same idea, when the keepalive breaks the connection because of an exception, we currently just break with that exception, so it's impossible to know that the break occurred as part of a keepalive. We could again prefix that with "Exception while sending a keepalive".